core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2554,6 +2557,19 @@ object SparkContext extends Logging {
     }
   }
 
+  /**
+   * Called to ensure that SparkContext is created or accessed only on the Driver.
+   *
+   * Throws an exception if a SparkContext is about to be created in executors.
+   */
+  private def assertOnDriver(): Unit = {
+    if (TaskContext.get != null) {
```

Comment:
@ueshin could you submit a follow-up PR to add a conf? In Spark 3.0, turn it off by default; in Spark 3.1 turn it on by default? Also add it to the migration guide?

Reply:
Does this actually affect any legitimate use case that would otherwise work? this should be more of a fail-fast for things that will already fail

Reply:
This is a user error. However, the users may use a Spark library in executors but the library calls SparkContext.getOrCreate. We should still let their workloads work if it worked in the previous releases.

Reply:
I see, but would that ever succeed, using the SparkContext that was created? I'm trying to work out what library would do this and not fail. Is it, perhaps, some boilerplate initialization code that gets executed on driver and executor, and it makes a SparkContext that is never actually used on the executor, but now it fails fast?

I get it, if that's the use case. A release note and/or hidden config to disable it might be an OK workaround.

Alternatively if arguing that this is not-uncommon, maybe we just don't do this at all, and revert it

Reply:
BTW, had been discussing this offline with @ueshin and @HyukjinKwon - one use case that previously (and admittedly unfortunately) relied on the ability to create `SparkContext`s in the executors is MLflow's `mlflow.pyfunc.spark_udf` API ([see docs & usage example](https://www.mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#mlflow.pyfunc.spark_udf)), which provides a unified interface for scoring models trained in arbitrary ML frameworks (tensorflow, scikit-learn, pyspark.ml) on Spark DataFrames as a pandas UDF. Most ML frameworks (e.g. sklearn) are single-node frameworks that can operate on pandas dataframes, so applying them via pandas UDF works well. For a pyspark.ml model to be applied via pandas UDF, we need to convert the input pandas series -> spark dataframe on the executors, which requires a SparkContext on the executors.

I'll dig to see if there's a way to keep the MLflow use case working with this change (e.g. pyspark.ml models may ultimately perform inference via spark UDF, so maybe we can somehow extract the underlying UDF from the model and return that from `mlflow.pyfunc.spark_udf`?), but otherwise agree that we may want to revert this change.

Reply:
I also sympathize with the notion that it's an antipattern to create a `SparkContext` on the executors, but not sure there's always a workaround for doing so

Reply:
Yep if that's the logic - that one might previously have harmlessly created a SparkContext that does not work and now it fails explicitly - then I'd say just revert it. It's just trying to fail-fast, but maybe that's not a good idea. If it affects mlflow, I'd guess it affects N other applications.

Reply:
I agree with having a config, even if the default is disable for a while, like during 3.x series.
I hope the config will be enabled by default eventually since I believe `SparkContext` is supposed to not be created in executors.

Reply:
This is pretty similar to the old `spark.driver.allowMultipleContexts`. It looks like working if you are lucky, but there are some surprising behaviors which are pretty hard to debug. For example, `SparkEnv.get` will be set to a new one when someone creates a SparkContext in an executor. (Search `SparkEnv.get` and you will find lots of codes running in executors call it).

I'm pretty surprised that nobody reports this issue in MLFlow. For myself, I used SparkContext in executors to run stress tests for codes running in driver, but I did hit multiple weird issues. I managed to workaround them by some hacks, such as setting Spark.env back. But it requires deep Spark knowledge. I believe many users won't be able to figure out the root cause without an explicit error added by this PR.

IMO, it's better to stop people from doing this, but we should also add a flag to give the users an option to turn it off, so that they have some time to migrate their workloads away from the fragile pattern. We can eventually remove the flag like what we did for `spark.driver.allowMultipleContexts`.

Reply:
@smurching how does mlflow create SparkContext? IIUC, it doesn't create SparkContext in the executor JVM, so it should not be impacted by this change.

Reply:
NVM. I thought the check was only added in JVM side. If we remove the pyspark side change, it should not impact MLflow. That's actually a case will work pretty well...

Reply:
So I'm +1 on blocking creating SparkContext in executors because it messes up the singleton `SparkEnv` (there may be others). But blocking it in Python process is probably not worth since it's working pretty well (unless in future we introduce some singleton objects in executor python processes).

Reply:
@zsxwing got it makes sense! Yeah, MLflow's spark UDF works pretty reliably AFAIK, at least for this use case :P. When scoring a pyspark.ml model in a pandas UDF, we construct a SparkContext via the standard SparkSession.builder.getOrCreate API ([see MLflow code](https://github.com/mlflow/mlflow/blob/86ea84f9e6d58b702f362ff512f61cff8ada021f/mlflow/spark.py#L520-L521))

Reply:
+1 for adding a config, disabling it in 3.0 and enabling it at 3.1.

Reply:
I submitted a PR to add configs #29278.
Could you take a look at it when you are available?

Reply:
I think it's a bit odds to have two separate configuration for Scala and Python. I don't think we will happen to allow this in PySpark specifically in the future even if it worked. I think we will discourage this behaviour anyway whether it is in Scala or Python, and deprecate/remove both configurations together eventually. In which case would we disable it in Python side but enable it in Python side?

Reply:
> In which case would we disable it in Scala side but enable it in Python side?

Like how MLflow uses it today. It's pretty hard to make it work correctly in Scala side. But it's working in Python today.

Reply:
I see but I guess we should discourage/deprecate/remove the behaviour in the end - am I correct? It's a bit weird to have two configuration to control the same behaviour (to end users). And I think it's a bit unlikely when it should be disabled in Scala but enabled in Python specifically. We could just enable it in both sides.

Reply:
I pushed a commit to use a single config.